server/src/test/java/org/elasticsearch/index/shard/IndexShardIT.java

```diff
@@ -771,20 +772,40 @@ public void testPendingRefreshWithIntervalChange() throws Exception {
         // wait for both to ensure we don't have in-flight operations
         updateSettingsLatch.await();
         refreshLatch.await();
-        // ensure no scheduled refresh to compete with the scheduleRefresh we are going to verify.
-        assertBusy(() -> {
-            for (ThreadPoolStats.Stats stat : indexService.getThreadPool().stats()) {
-                if (stat.getName().equals(ThreadPool.Names.REFRESH) && (stat.getQueue() > 0 || stat.getActive() > 0)) {
-                    throw new AssertionError(); // cause assert busy to retry
-                }
-            }
-        });
+        // We need to ensure a `scheduledRefresh` triggered by the internal refresh setting update is executed before we index a new doc;
+        // otherwise, it will compete to call `Engine#maybeRefresh` with the `scheduledRefresh` that we are going to verify.
+        ensureNoPendingScheduledRefresh(indexService.getThreadPool());
         client().prepareIndex("test", "test", "2").setSource("{\"foo\" : \"bar\"}", XContentType.JSON).get();
         assertTrue(shard.scheduledRefresh());
         assertTrue(shard.isSearchIdle());
         assertHitCount(client().prepareSearch().get(), 3);
     }
 
+    private void ensureNoPendingScheduledRefresh(ThreadPool threadPool) throws Exception {
+        // We can make sure that all scheduled refresh tasks are done by submitting some tasks (> maximumPoolSize so at least one task
+        // will be in the queue) to the refresh thread pool then wait until all of them are completed. Note that using ThreadPoolStats
+        // alone is not enough as both queue and active count can be 0 when ThreadPoolExecutor(see ThreadPoolExecutor#runWorker) just
+        // takes a task out the queue but before marking it active (i.e., lock the corresponding worker).
+        ThreadPoolExecutor refreshThreadPoolExecutor = (ThreadPoolExecutor) threadPool.executor(ThreadPool.Names.REFRESH);
+        int numTasks = refreshThreadPoolExecutor.getMaximumPoolSize() + 1;
+        CountDownLatch latch = new CountDownLatch(1);
+        for (int i = 0; i < numTasks; i++) {
+            refreshThreadPoolExecutor.execute(() -> {
+                try {
+                    latch.await();
```

Comment:
I'm not sure why this would work. Would we not rather have to submit `refreshThreadPoolExecutor.getMaximumPoolSize()` number of tasks and use a `CyclicBarrier` with `refreshThreadPoolExecutor.getMaximumPoolSize() + 1` permits to make sure that there is no worker that might still be in the process of picking up the settings change refresh task?

Reply:
It works due to the `assertBusy` at the end but yours's much simpler. I adopted it in  https://github.com/elastic/elasticsearch/pull/45025/commits/04774203852c10111c111135910a595a19d46ef5.